genesis/engine/solvers/rigid/gjk_decomp.py

```diff
@@ -0,0 +1,3394 @@
+                    k_1, k_2, k_3 = 1, 2, 3
+                elif i == 2:
+                    k_1, k_2, k_3 = 0, 2, 3
+                elif i == 3:
+                    k_1, k_2, k_3 = 0, 1, 3
+
+                _lambda2d, flag2d = self.func_simple_gjk_subdistance_2d(i_b, k_1, k_2, k_3)
+
+                if failed_3d:
+                    if flag2d:
+                        closest_point = self.func_simplex_vertex_linear_comb(i_b, 2, k_1, k_2, k_3, 0, _lambda2d, 3)
+                        d = closest_point.dot(closest_point)
+                        if d < dmin:
+                            dmin = d
+                            _lambda.fill(0.0)
+                            _lambda[k_1] = _lambda2d[0]
+                            _lambda[k_2] = _lambda2d[1]
+                            _lambda[k_3] = _lambda2d[2]
+                else:
+                    if flag2d:
+                        _lambda = _lambda2d
+                    flag = flag2d
+
+        if (flag == 0) or n == 2:
+            failed_3d = n == 4
+            failed_2d = n == 3
+
+            num_iter = 1
+            if failed_3d:
+                # Iterate through 6 edges of the tetrahedron
+                num_iter = 6
+            elif failed_2d:
+                # Iterate through 3 edges of the triangle
+                num_iter = 3
+
+            for i in range(num_iter):
+                k_1, k_2 = 0, 1
+                if i == 1:
+                    k_1, k_2 = 0, 2
+                elif i == 2:
+                    k_1, k_2 = 1, 2
+                elif i == 3:
+                    k_1, k_2 = 0, 3
+                elif i == 4:
+                    k_1, k_2 = 1, 3
+                elif i == 5:
+                    k_1, k_2 = 2, 3
+
+                _lambda1d = self.func_simple_gjk_subdistance_1d(i_b, k_1, k_2)
+
+                if failed_3d or failed_2d:
+                    closest_point = self.func_simplex_vertex_linear_comb(i_b, 2, k_1, k_2, 0, 0, _lambda1d, 2)
+                    d = closest_point.dot(closest_point)
+                    if d < dmin:
+                        dmin = d
+                        _lambda.fill(0.0)
+                        _lambda[k_1] = _lambda1d[0]
+                        _lambda[k_2] = _lambda1d[1]
+                else:
+                    _lambda = _lambda1d
+
+        return _lambda
+
+    @ti.func
+    def func_simple_gjk_subdistance_3d(self, i_b, i_s1, i_s2, i_s3, i_s4):
+
+        flag = 0
+        _lambda = gs.ti_vec4(0, 0, 0, 0)
+
+        # Simplex vertices
+        s1 = self.gjk_simplex_vertex[i_b, i_s1].mink
+        s2 = self.gjk_simplex_vertex[i_b, i_s2].mink
+        s3 = self.gjk_simplex_vertex[i_b, i_s3].mink
+        s4 = self.gjk_simplex_vertex[i_b, i_s4].mink
+
+        # Compute the cofactors to find det(M),
+        # which corresponds to the signed volume of the tetrahedron
+        Cs = ti.math.vec4(0.0, 0.0, 0.0, 0.0)
+        for i in range(4):
+            v1, v2, v3 = s2, s3, s4
```

Comment:
What about `v1, v2, v3 = ti.static(s2, s3, s4)` (same for the other lines) ? This may improve performance a bit.
See https://docs.taichi-lang.org/docs/syntax_sugars.

Reply:
My bad then!